### PR TITLE
Get modules information using proc_pidinfo() on macOS

### LINKED ISSUE #42888: Debugger not working on macOS 10.15.x when vmmap hides full path to process modules
Hi! We found that sometimes `vmmap` that you use for `CreateProcessModules` (https://github.com/dotnet/runtime/blob/d52ee3ada3da7093dd0bdc08bf2c816d7dd2a646/src/coreclr/src/pal/src/thread/process.cpp#L2860), can return paths in a "secure" manner, replacing some parts with *.

For example, on my Mac mini with macOS 10.15.7, when I run dotnet from the `~/Downloads` directory, I see paths like:
```__TEXT                        109fc4000-109fd1000        [   52K    52K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/dotnet
__TEXT                        109ff1000-10a044000        [  332K   332K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libhostfxr.dylib
__TEXT                        10a04c000-10a091000        [  276K   276K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libhostpolicy.dylib
__TEXT                        10a09a000-10a526000        [ 4656K  4656K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libcoreclr.dylib
__TEXT                        10a526000-10a527000        [    4K     4K     4K     0K] r--/rwx SM=COW          /Users/USER/*/libcoreclr.dylib
__TEXT                        10a527000-10a5fa000        [  844K   844K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libcoreclr.dylib
__TEXT                        10abb0000-10add6000        [ 2200K  2200K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libclrjit.dylib
__TEXT                        10ae6b000-10ae75000        [   40K    40K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/System.Native.dylib
__TEXT                        10ae79000-10ae82000        [   36K    36K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/System.Globalization.Native.dylib
__TEXT                        112036000-1120c8000        [  584K   568K     0K     0K] r-x/r-x SM=COW          /usr/lib/dyld
```

In this case, we get `CORDBG_E_DEBUG_COMPONENT_MISSING` on debugger initialization (https://github.com/dotnet/runtime/blob/d52ee3ada3da7093dd0bdc08bf2c816d7dd2a646/src/coreclr/src/dlls/dbgshim/dbgshim.cpp#L386) because the `dbiPath` is `/Users/USER/*/libmscordbi.dylib` and `LoadLibraryA` cannot find it.

Do you have any objection to using libproc in `CreateProcessModules`? Looks like iterating over regions using this API gives normal paths, so I'm going to submit a PR. cc @mikem8361 @janvorli @jkotas

### PR DESCRIPTION
**Pull request by k15tfu**
Previously CreateProcessModules() parsed the output of vmmap command to get
modules addresses/paths on macOS, but on some Macs with latest macOS 10.15.6
vmmap hides full paths to some process modules (.dylibs in non-system folders),
replacing some parts with *:
```
__TEXT                 000000010d8bd000-000000010ddce000 [ 5188K  5188K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libcoreclr.dylib
```

In particular, it breaks the debugger functionality due to invalid path
`/Users/USER/*/libmscordbi.dylib`, and error code
CORDBG_E_DEBUG_COMPONENT_MISSING is returned.

Now we get modules information by iterating over regions using proc_pidinfo()
in CreateProcessModules().

Fixes #42888

### CODE REVIEW

File: src/coreclr/src/pal/src/thread/process.cpp [hunk 1]
```diff
@@ -2851,65 +2852,63 @@ CreateProcessModules(
...
-    _ASSERTE(chars > 0 && chars <= sizeof(vmmapCommand));
+    // NOTE: Sometimes vmmap hides full paths to some process modules (.dylibs in non-system folders), causing debugger not to work.
+    // __TEXT                 000000010d8bd000-000000010ddce000 [ 5188K  5188K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libcoreclr.dylib
+    // So now we get modules information by iterating over regions using proc_pidinfo().  See dotnet/runtime#42888.
+    int count = 0;
 
-    FILE *vmmapFile = popen(vmmapCommand, "r");
-    if (vmmapFile == NULL)
+    uint64_t addr = 0;
+    for (;;)
```
Comment (am11) [inline src/coreclr/src/pal/src/thread/process.cpp:2862]: nit: 
```suggestion
    while (true)
```

File: src/coreclr/src/pal/src/thread/process.cpp [hunk 2]
```diff
@@ -2851,65 +2852,63 @@ CreateProcessModules(
...
+                dup = true;
+                break;
             }
+        }
 
-            if (!dup)
+        if (!dup)
+        {
+            int cbModuleName = strlen(moduleName) + 1;
+            ProcessModules *entry = (ProcessModules *)InternalMalloc(sizeof(ProcessModules) + cbModuleName);
```
Comment (am11) [inline src/coreclr/src/pal/src/thread/process.cpp:2892]: Does it need a pairing `InternalDelete(entry)`?

Comment (k15tfu) [reply src/coreclr/src/pal/src/thread/process.cpp:2892]: Not here, it's already done in DestroyProcessModules().

**Comment by mikem8361** [comment]
The changes look good.  I'm sure the changes will fix your issues, but I don't know enough about proc_pidinfo yet to know if it will break our existing scenarios (other MacOS versions, devices etc.).  I want to wait a little while to approve and merge until 5.0 is finished and out the door.

And maybe our CI tests will get more stable though you will have to rebase on the latest master to get any fixes.

**Comment by k15tfu** [comment]
@mikem8361 Hi! Glad to hear this will be included in 5.0 =) I have just rebased against TOT to see if tests pass now.

And would it also be possible to backport this to 3.1?

**Comment by mikem8361** [comment]
Actually it won't make it into 5.0.  It is too risky change right now (weeks before we release 5.0). We will have to see about 3.1/5.0 service releases.

**Comment by k15tfu** [comment]
@mikem8361 Oops, I misread this. 5.1 still sounds good :)

**Review by mikem8361 — APPROVED** [review]

**Comment by k15tfu** [comment]
Hi @mikem8361. Thanks for approving this. So how I can help to get this fixed in 3.1/5.0 service releases?

**Comment by mikem8361** [comment]
It may be awhile before anything goes into 5.0 servicing. We are still trying to finish 5.0.  For 3.1, the it may also be some time before I can go though all our internal servicing processes. The servicing bar is pretty high.

/cc: @Tom McDonald<mailto:Thomas.McDonald@microsoft.com>

**Comment by k15tfu** [comment]
@mikem8361 Kindly reminder.

**Comment by stevo-knievo** [comment]
**+1**
for getting this fix in 3.1/5.0 service releases, please.

**Pull request merged**